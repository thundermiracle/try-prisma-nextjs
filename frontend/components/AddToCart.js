import React from "react";
import PropTypes from "prop-types";

import { Mutation } from "react-apollo";
import gql from "graphql-tag";

const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ id }) => {
  return (
    <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ id }}>
      {(addToCart) => <button onClick={addToCart}>Add to CartðŸ›’</button>}
    </Mutation>
  );
};

AddToCart.propTypes = {};

export default AddToCart;
