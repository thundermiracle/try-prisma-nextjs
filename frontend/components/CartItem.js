import React from "react";
import PropTypes from "prop-types";

import styled from "styled-components";
import formatMoney from "../lib/formatMoney";
import RemoveFromCart from "./RemoveFromCart";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ data: { id, item, quantity } }) => {
  const { image, title, price } = item;

  return (
    <CartItemStyles>
      <img width="100" height="70" src={image} alt={title} />
      <div className="cart-item-details">
        <h3>{title}</h3>
        <p>
          {formatMoney(price * quantity)}
          {"-"}
          <em>
            {quantity} &times; {formatMoney(price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={id} />
    </CartItemStyles>
  );
};

CartItem.propTypes = {};

export default CartItem;
