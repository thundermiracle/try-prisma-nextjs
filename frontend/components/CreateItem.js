import React from "react";
import PropTypes from "prop-types";

import Router from "next/router";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import formatMoney from "../lib/formatMoney";
import ErrorMessage from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      image: $image
      largeImage: $largeImage
      price: $price
    ) {
      id
    }
  }
`;

const CreateItem = (props) => {
  const [state, setState] = React.useState({
    title: "",
    description: "",
    image: "",
    largeImage: "",
    price: 0,
  });

  const handleChange = (e) => {
    const { name, type, value } = e.target;
    const val = type === "number" ? parseFloat(value) || 0 : value;

    setState({
      ...state,
      [name]: val,
    });
  };

  const uploadFile = async (e) => {
    const files = e.target.files;
    const data = new FormData();
    data.append("file", files[0]);
    data.append("upload_preset", "sickfits");

    const res = await fetch(
      "https://api.cloudinary.com/v1_1/thundermiracle/image/upload",
      {
        method: "POST",
        body: data,
      },
    );

    const file = await res.json();
    console.log("image uploaded");
    setState({
      ...state,
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  };

  return (
    <Mutation mutation={CREATE_ITEM_MUTATION} variables={state}>
      {(createItem, { loading, error, called, data }) => {
        return (
          <Form
            onSubmit={async (e) => {
              e.preventDefault();
              const res = await createItem();

              Router.push({
                pathname: "/item",
                query: {
                  id: res.data.createItem.id,
                },
              });
            }}
          >
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  value={state.file}
                  onChange={uploadFile}
                />
                {state.image && <img src={state.image} alt="Upload preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={state.title}
                  onChange={handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={state.price}
                  onChange={handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a Description"
                  required
                  value={state.description}
                  onChange={handleChange}
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        );
      }}
    </Mutation>
  );
};

CreateItem.propTypes = {};

export { CreateItem as default, CREATE_ITEM_MUTATION };
