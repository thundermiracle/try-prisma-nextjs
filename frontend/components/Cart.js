import React from "react";
import PropTypes from "prop-types";

import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import User from "./User";
import CartItem from "./CartItem";
import CartStyles from "./styles/CartStyles";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import SickButton from "./styles/SickButton";

import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
import TakeMyMoney from "./TakeMyMoney";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`;

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => (
    <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>
  ),
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
});

const Cart = (props) => {
  return (
    <Composed>
      {({ user, toggleCart, localState }) => {
        if (!user.data || !user.data.me) {
          return null;
        }
        const me = user.data.me;

        return (
          <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton title="close" onClick={toggleCart}>
                &times;
              </CloseButton>
              <Supreme>{me.name}'s Cart</Supreme>
              <p>You Have {me.cart.length} Items in your cart.</p>
            </header>
            <ul style={{ overflow: "auto" }}>
              {me.cart.map((cartItem) => (
                <CartItem key={cartItem.id} data={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              {me.cart.length > 0 ? (
                <TakeMyMoney>
                  <SickButton>Checkout</SickButton>
                </TakeMyMoney>
              ) : null}
            </footer>
          </CartStyles>
        );
      }}
    </Composed>
  );
};

Cart.propTypes = {};

export default Cart;
