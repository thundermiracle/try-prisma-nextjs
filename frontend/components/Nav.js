import Link from "next/link";
import { Mutation } from "react-apollo";
import { TOGGLE_CART_MUTATION } from "./Cart";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import ErrorMessage from "./ErrorMessage";
import Signout from "./Signout";
import CartCount from "./CartCount";

const Nav = () => (
  <NavStyles>
    <Link href="/items">
      <a>Shop</a>
    </Link>
    <User>
      {({ data, error, loading }) => {
        if (loading) return null;
        if (error) return <ErrorMessage error={error} />;
        if (!data || !data.me)
          return (
            <Link href="/signup">
              <a>Signin</a>
            </Link>
          );

        return (
          <>
            <Link href="/sell">
              <a>Sell</a>
            </Link>
            <Link href="/orders">
              <a>Orders</a>
            </Link>
            <Link href="/me">
              <a>Account</a>
            </Link>
            <Signout />
            <Mutation mutation={TOGGLE_CART_MUTATION}>
              {(toggleCart) => (
                <button onClick={toggleCart}>
                  My Cart
                  <CartCount
                    count={data.me.cart.reduce((prevCount, { quantity }) => {
                      return prevCount + quantity;
                    }, 0)}
                  />
                </button>
              )}
            </Mutation>
          </>
        );
      }}
    </User>
  </NavStyles>
);

export default Nav;
