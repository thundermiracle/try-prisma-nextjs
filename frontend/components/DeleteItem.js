import React from "react";
import PropTypes from "prop-types";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

const DeleteItem = ({ id, children }) => {
  const update = (cache, payload) => {
    // manually update the cache on client
    // 1. read the cache
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });

    // 2. filter the deleted item out of the page
    const items = data.items.filter((item) => item.id !== id);

    // 3. put the items back
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data: { ...data, items } });
  };

  return (
    <Mutation
      mutation={DELETE_ITEM_MUTATION}
      variables={{ id }}
      update={update}
    >
      {(deleteItem, { error, loading }) => (
        <button
          onClick={() => {
            if (confirm("Are you sure?")) {
              deleteItem().catch((err) => alert(err.message));
            }
          }}
        >
          {children}
        </button>
      )}
    </Mutation>
  );
};

DeleteItem.propTypes = {};

export default DeleteItem;
