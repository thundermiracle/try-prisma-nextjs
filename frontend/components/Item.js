import React from 'react'
import PropTypes from 'prop-types'
import Link from 'next/link';

import Title from './styles/Title';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteItem from './DeleteItem';

const Item = ({ item: { id, title, price, description, image } }) => {
  return (
    <ItemStyles>
      {image && <img src={image} alt={title} />}
      <Title>
        <Link
          href={{
            pathname: "/item",
            query: { id }
          }}
        >
          <a>{title}</a>
        </Link>
      </Title>
      <PriceTag>{formatMoney(price)}</PriceTag>
      <p>{description}</p>

      <div className="buttonList">
        <Link href={{ pathname: "update", query: { id } }}>
          <button><a>Edit</a></button>
        </Link>
        <button>Add to Cart</button>
        <DeleteItem id={id}>Delete this item</DeleteItem>
      </div>
    </ItemStyles>
  );
};

Item.propTypes = {
  item: PropTypes.object.isRequired,
}

export default Item
