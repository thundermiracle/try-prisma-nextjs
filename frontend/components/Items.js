import React from "react";
import PropTypes from "prop-types";

import { Query } from "react-apollo";
import gql from "graphql-tag";
import style from "styled-components";
import ErrorMessage from "./ErrorMessage";
import Item from "./Item";
import Pagination from './Pagination'
import { perPage } from "../config";

export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
    items(skip: $skip, first: $first, orderBy: createdAt_DESC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

const Center = style.div`
  text-align: center;
`;

const ItemList = style.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 60px;
  max-width: ${(props) => props.theme.maxWidth};
  margin: 0 auto;
`;

const Items = ({ page }) => {
  return (
    <Center>
      <Pagination page={page} />
      <Query
        query={ALL_ITEMS_QUERY}
        fetchPolicy="network-only"
        variables={{
          skip: (page - 1) * perPage,
          first: perPage
        }
        }>
        {({ data, error, loading }) => {
          if (loading) return <p>loading...</p>;
          if (error) return <ErrorMessage error={error} />;

          return (
            <ItemList>
              {data.items.map((item) => (
                <Item item={item} key={item.id} />
              ))}
            </ItemList>
          );
        }}
      </Query>
      <Pagination page={page} />
    </Center>
  );
};

Items.propTypes = {};

export default Items;
