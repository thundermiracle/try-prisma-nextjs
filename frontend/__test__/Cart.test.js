import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import { MockedProvider } from "@apollo/react-testing";

import {
  fakeCartItem,
  fakeUser,
  waitForComponentToPaint,
} from "../lib/testUtils";
import { CURRENT_USER_QUERY } from "../components/User";
import Cart, { LOCAL_STATE_QUERY } from "../components/Cart";

const me = fakeUser();
const mocks = [
  // local state query
  {
    request: {
      query: LOCAL_STATE_QUERY,
    },
    result: {
      data: {
        cartOpen: true,
      },
    },
  },

  // current user query mock
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      data: {
        me: {
          __typename: "User",
          ...me,
          cart: [fakeCartItem()],
        },
      },
    },
  },
];

describe("<Cart />", () => {
  it("renders correctly", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Cart />
      </MockedProvider>,
    );

    expect(toJSON(wrapper.find("header"))).toMatchSnapshot();

    // CANNOT get localState correctly
    // await waitForComponentToPaint(wrapper);
    // expect(wrapper.find("CartItem")).toHaveLength(1);
  });
});
