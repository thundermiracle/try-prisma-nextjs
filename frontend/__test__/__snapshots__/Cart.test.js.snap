// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders correctly 1`] = `null`;
